### DISCUSSION TITLE
ConcurrentT.Stage: Why does it not spawn a go routine by itself?

### DISCUSSION THREAD

**Discussion by matthiasgeihs** [hyperledger-labs/go-perun#28] [Q&A]
@RmbRT During a code review, I saw that method `Stage` of `ConcurrentT` is usually run in a go routine:
```
go ct.Stage(...)
...
ct.Wait(...)
```

I wonder: Why don't we provide a method that spawns a go routine by default?
```
ct.StartStage(...)
...
ct.Wait(...)
```
Is there at all a need for a method that runs in the same routine?

**Comment by ggwpez** [comment]
We do have code where we already have a go-routine spawned, eg:  
```go
go func() {
    // Other Code
    ct.Stage(…)
    // Other Code
}()
```
It also makes sense to have it when using `ctxtest.AssertTerminates`.  

I think changing it to implicit go-routine spawning would also make the usage less obvious to the user. Since `Stage(N)` is often used in for loops, a user always has to capture the variables via copy instead of reference since they would otherwise change.  
That is only necessary when running in a separate go-routine.

The following code would be wrong without raising much suspicion:
```go
for i := 0;; i++ {
    ct.StartStage(/* use i */)
}
```
Where as the following pattern is often found in the code:
```go
for i := 0;; i++ {
    i := i
    go ct.Stage(/* use i */)
}
```
